File: 백준/Platinum/11266. 단절점/단절점.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(int(1e6))

def dfs(now, is_root):
    global cnt
    cnt += 1
    visited[now] = cnt
    ret = visited[now]
    child = 0
    
    for next in graph[now]:
        if not visited[next]:
            child += 1
            low = dfs(next, False)
            ret = min(ret, low)
            
            if not is_root and low >= visited[now]:
                cut_vertex[now] = True
        else:
            ret = min(ret, visited[next])
    
    if is_root and child >= 2:
        cut_vertex[now] = True
    
    return ret
    


v,e = map(int, input().split())
graph = [[] for _ in range(v+1)]

for _ in range(e):
    a,b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    
cnt = 0
visited = [0]*(v+1)
cut_vertex = [False]*(v+1)

for i in range(1, v+1):
    if not visited[i]:
        dfs(i, True)

res = []
for i in range(1, v+1):
    if cut_vertex[i]:
        res.append(i)

print(len(res))
print(*res)

File: 백준/Platinum/11266. 단절점/단절점.cc
#include <bits/stdc++.h>
#define ll long long
//#define int long long
const int inf = 1e9;

using namespace std;

int n,m;
int cnt = 1;
int order[10101];
int cut[10101];
vector<int> graph[10101];

int dfs(int now){
    order[now] = cnt++;

    int ret = order[now];
    int child = 0;

    for (auto next : graph[now]){
        if (!order[next]){
            int low = dfs(next);
            ret = min(ret, low);
            child++;

            if (order[now] != 1 && low >= order[now]) cut[now] = 1;
        } else {
            ret = min(ret, order[next]);
        }
    }
    
    if (order[now] == 1 && child >= 2) cut[now] = 1;
    
    return ret;
}


int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> m;
    for (int i = 0; i < m; i++){
        int u,v; cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    for (int i = 1; i <= n; i++){
        if (!order[i]) {
            cnt = 1;
            dfs(i);
        }
    }

    vector<int> ans;

    for (int i = 1; i <= n; i++){
        if (cut[i]) ans.push_back(i);
    }

    cout << ans.size() << '\n';
    for (auto v : ans) cout << v << ' ';
    
    return 0;
}